keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,788 @@
+  Each pod is placed in a different user namespace and has a different and
+  non-overlapping ID mappings. This mode is intended for stateless pods, i.e.
+  pods using only ephemeral volumes like `configMap,` `downwardAPI`, `secret`,
+  `projected` and `emptyDir`. This mode not only provides host-to-pod isolation
+  but also  pod-to-pod isolation as each pod has a different range of effective
+  IDs in the host.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to give any extra privilege on the host.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to share the host user namespace
+if they need a feature only available in such user namespace.
+
+### Notes/Constraints/Caveats
+
+#### Volumes Support
+
+The Linux kernel uses the effective user and group IDs (the ones the host) to
+check the file access permissions. Since with user namespaces IDs are mapped to
+a different value on the host, this could cause issues accessing volumes if the
+pod is run with a different mapping, i.e. the effective user and group IDs on
+the host change.
+
+This proposal supports volume without changing the user and group IDs and leaves
+that problem to the user to manage. Future Linux kernel features such as shiftfs
+could allow to different pods to see a volume with its own IDs but it is out of
+scope of this proposal. Among the possible future kernel solutions, we can list:
+
+- [shiftfs: uid/gid shifting filesystem](https://lwn.net/Articles/757650/)
+- [A new API for mounting filesystems](https://lwn.net/Articles/753473/)
+- [user_namespace: introduce fsid mappings](https://lwn.net/Articles/812221/)
+
+In regard to this proposal, volumes can be divided in ephemeral and non-ephemeral.
+
+Ephemeral volumes are associated to a **single** pod and their lifecyle is
+dependent on that pod. These are `configMap`, `secret`, `emptyDir`,
+`downwardAPI`, etc. These kind of volumes are easy to handle as they are not
+shared by different pods and hence all the process accessig those volumes have
+the same effective user and group IDs. Kubelet creates the files for those
+volumes and it can easily set the file ownership too.
+
+Non-ephemeral volumes more difficult to support since they can be persistent and
+also can be shared by multiple pods. This proposal supports volumes with two
+different strategies:
+- The `Cluster` makes it easier for pods to share files using volumes when those
+  don't have access permissions for `others` because the effective user and
+  group IDs on the host are the same for all the pods.
+- The semantics of semantics of `fsGroup` are respected, if it's specified it's
+  assumed to be the correct GID in the host and an 1-to-1 mapping entry for the
+  `fsGroup` is added to the GID mappings for the pod.
```

Comment:
No, it's independent and also added for `Pod` mode.
The motivation is that a pod running in `Pod` mode is able to access the files in a volume if they are owned by `fsGroup`.

Reply:
Ok, so `fsGroup: X` means a line like this is added: `X, X, 1` to the mapping file?

This might have security considerations to consider too, right? As long as we don't forget or are tracking them in some way, I'm fine :)

Reply:
Marking as solved, but feel free to re-open if I'm missing something

Reply:
You're right, that line is added. `fsGroup` can already be controlled by the PSP, so users can control what's allowed there. We should make it clear on the documentation that this special entry in the mapping is added for this field.